docs/language/learn-ql/cpp/value-numbering-hash-cons.rst

```diff
@@ -0,0 +1,104 @@
+Hash consing and value numbering
+=================================================
+Overview
+--------
+In C and C++ QL databases, each node in the abstract syntax tree is represented by a separate object. This allows both analysis and results display to refer to specific appearances of a piece of syntax. However, it is frequently useful to determine whether two expressions are equivalent, either syntactically or semantically.
```

Comment:
_This allows both *analysis* and results display to refer to specific appearances of a piece of syntax._ 

When you say **analysis**, do you mean **extraction**? The CodeQL CLI is going to use an subcommand `analyze` for running one or more queries on a QL database, so it might be better to explicitly say extraction here, if that's what you mean.

Reply:
I also think that slightly more detailed explanations of hash consing and value numbering would be helpful here. Perhaps with links to wikipedia pages?

Reply:
> When you say analysis, do you mean extraction?

No, I meant the stage where the queries run. This could probably be worded better, though.